cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        throw new IOException(e);
+      }
+    }
+  }
+
+  /**
+   * Wrapper around a {@link StreamSizeSchedule} which will run a program whenever it receives enough
+   * data from a stream, via notifications.
+   */
+  private final class StreamSizeScheduleTask {
+    private final Id.Program programId;
+    private final SchedulableProgramType programType;
+    private StreamSizeSchedule streamSizeSchedule;
+
+    private long baseSize;
+    private long baseTs;
+    private AtomicBoolean active;
+
+    private StreamSizeScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule) {
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+    }
+
+    public void startSchedule(long baseSize, long baseTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with baseSize {}, baseTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), baseSize, baseTs, active, persist);
+      this.baseSize = baseSize;
+      this.baseTs = baseTs;
+      this.active = new AtomicBoolean(active);
+    }
+
+    public boolean isActive() {
+      return active.get();
+    }
+
+    public void received(StreamSizeNotification notification) {
+      if (!active.get()) {
+        return;
+      }
+      long pastRunSize;
+      long pastRunTs;
+      StreamSizeSchedule currentSchedule;
+      synchronized (this) {
+        if (notification.getSize() < baseSize) {
+          // This can happen if no notification is received for a stream for some time, and we poll the stream events
+          // size using metrics, and some metric events have hit their TTL. In that case it's impossible to know
+          // how much data was ingested, and how much data has hit the TTL. Resetting the base attributes is the best
+          // we can do.
+          baseSize = notification.getSize();
+          baseTs = notification.getTimestamp();
+          return;
+        }
+        currentSchedule = streamSizeSchedule;
+        if (notification.getSize() < baseSize + toBytes(currentSchedule.getDataTriggerMB())) {
+          return;
+        }
+
+        // Update the baseSize as soon as possible to avoid races
+        pastRunSize = baseSize;
+        pastRunTs = baseTs;
+        baseSize = notification.getSize();
+        baseTs = notification.getTimestamp();
+        LOG.debug("Base size and ts updated to {}, {} for streamSizeSchedule {}",
+                  baseSize, baseTs, currentSchedule);
+      }
+
+      Arguments args = new BasicArguments(ImmutableMap.of(
+        ProgramOptionConstants.SCHEDULE_NAME, currentSchedule.getName(),
+        ProgramOptionConstants.LOGICAL_START_TIME, Long.toString(baseTs),
+        ProgramOptionConstants.RUN_DATA_SIZE, Long.toString(baseSize),
+        ProgramOptionConstants.PAST_RUN_LOGICAL_START_TIME, Long.toString(pastRunTs),
```

Comment:
Is this always the logical start time of the last run? Or or is this the time that corresponds to the base size that was used to compute the amount of new data? Due to the special case above (and also if a run fails), it could be that there was no run at that time.